File: hard/text_to_number.hs
import System.Environment (getArgs)
import Data.List (elemIndex)
import Data.Maybe (fromJust, isJust)

d0 :: [String]
d0 = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
d1 :: [String]
d1 = ["ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen",
      "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
v1 :: [Int]
v1 = [10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 30, 40, 50, 60, 70, 80, 90]

ttn             :: Int -> Int -> Int -> [String] -> Int
ttn r h q []     = r + q + h
ttn r h q (x:xs) | x == "zero"     = ttn r h 0 xs
                 | isJust d0i      = ttn r (fromJust d0i) q xs
                 | isJust d1i      = ttn r h (v1 !! fromJust d1i + q) xs
                 | x == "hundred"  = ttn r 0 (h*100) xs
                 | x == "thousand" = ttn (r+(h+q)*1000) 0 0 xs
                 | x == "million"  = ttn ((h+q)*1000000) 0 0 xs
                 where d0i = elemIndex x d0
                       d1i = elemIndex x d1

texttonumber       :: [String] -> Int
texttonumber []     = 0
texttonumber (x:xs) | x == "negative" = - texttonumber xs
                    | otherwise       = ttn 0 0 0 (x:xs)

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (show . texttonumber . words) $ lines input

File: hard/text_to_number.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

const dn, dz = "negative", "zero"

var (
	d0, d1, d2 map[string]int
	d3         []string
)

func init() {
	d0 = map[string]int{"zero": 0, "one": 1, "two": 2,
		"three": 3, "four": 4, "five": 5, "six": 6,
		"seven": 7, "eight": 8, "nine": 9}
	d1 = map[string]int{"ten": 10, "eleven": 11, "twelve": 12,
		"thirteen": 13, "fourteen": 14, "fifteen": 15, "sixteen": 16,
		"seventeen": 17, "eighteen": 18, "nineteen": 19}
	d2 = map[string]int{"twenty": 20, "thirty": 30, "forty": 40,
		"fifty": 50, "sixty": 60, "seventy": 70, "eighty": 80,
		"ninety": 90}
	d3 = []string{"hundred", "thousand", "million"}
}

func textToNumber(s string) int {
	t := strings.Fields(s)
	var n, h, q, r int
	if t[0] == dn {
		n = -1
		t = t[1:]
	} else {
		n = 1
	}
	for len(t) > 0 {
		if t[0] == dz {
			q = 0
		} else if v, f := d0[t[0]]; f {
			h = v
		} else if v, f := d1[t[0]]; f {
			q += v
		} else if v, f := d2[t[0]]; f {
			q += v
		} else if t[0] == d3[0] {
			q, h = h*100, 0
		} else if t[0] == d3[1] {
			r, h, q = r+(h+q)*1000, 0, 0
		} else if t[0] == d3[2] {
			r, h, q = (h+q)*1000000, 0, 0
		}
		t = t[1:]
	}
	return n * (r + q + h)
}

func main() {
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()

	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		fmt.Println(textToNumber(scanner.Text()))
	}
}
